Make diagnostics only on test files

lyy011lyy opened the pull request:
Currently, ts-jest is checking the type errors for source codes and all
their dependencies if "enableTsDiagnostics" is turned ON. This makes
it very slow and I don't think that's what we want since the src code
type errors will be detected during the build time.

What we need should be a way to report the type errors only inside the
test files, because usually test code will be excluded when build.

Add logic to check whether the filePath is inside one of the
root's pathes to achieve that only check the type errors inside test
code.

huafu commented on src/preprocess.ts:
```diff
@@ -42,7 +42,10 @@ export default function preprocess(
-  if (tsJestConfig.enableTsDiagnostics) {
+  if (
+    tsJestConfig.enableTsDiagnostics &&
+    jestConfig.roots.find(root => filePath.indexOf(root) > -1)
```
I believe this will work only if you choose a `rootDir` containing ONLY your test files. I think what you wanted here is something using `testMatch` (or `testRegex`, whether is set) and `testPathIgnorePatterns`. See https://jestjs.io/docs/en/configuration.html#testmatch-array-string.

Or did I misunderstood the purpose of this PR?

lyy011lyy commented:
Hi @huafu :
Can I ask that is there any progress on this request? This is important to our project and we really need this small change.
If there is any other change that I can make, please feel free to reply and I will try my best.
Your help would be appreciated!

lyy011lyy commented:
@huafu Thanks for the reply.
I see, the one you mentioned is testMatch.
So user should only specify either testMatch(array) or testRegex. 
Do you mean that I should add the logic to check both of them ?

Thanks!

huafu commented:
@lyy011lyy yup you'll have to add the logic. Badly jest doesn't export it from its lib, but you can have a look how it's done internally there:

https://github.com/facebook/jest/blob/79c9e22c1309eb92e41f5cad74c767c6a91f68d9/packages/jest-runtime/src/script_transformer.js#L484-L502

huafu commented:
ok here you go https://github.com/facebook/jest/blob/9afeb9c757e4b92dbf670e66545dce3df5c407c4/packages/jest-runtime/src/should_instrument.js#L38-L48

Thanks for doing the change!

lyy011lyy commented:
@huafu  cool! 
Thanks for your help! Will upload the change asap.

lyy011lyy commented:
@huafu  Emm.... this is a much better solution I think !
Thanks and I will change it in this way!

huafu commented on README.md:
```diff
@@ -12,35 +12,35 @@
+    - [`const enum` is not supported](#const-enum-is-not-supported)
+  - [How to Contribute](#how-to-contribute)
+    - [Quickstart to run tests (only if you're working on this package)](#quickstart-to-run-tests-only-if-youre-working-on-this-package)
+  - [License](#license)
```
Can you actually undo the spaces change here please?

lyy011lyy replied:
Done, it's my markDown plugin to update the tableOfContent on save...
Sorry about this.

huafu commented on README.md:
```diff
@@ -250,6 +251,17 @@ If you want to enable Syntactic & Semantic TypeScript error reporting you can en
+      }
+    }
+  }
+}
```
Thanks for adding this!

lyy011lyy commented:
Hi @huafu :
I just upload the change and also change the document.
Thanks

lyy011lyy commented:
@huafu 
Sure, I'm adding test for it.

huafu commented:
@lyy011lyy thanks for trying. Please do not force push every time, like this I can help see what's going on to help you.

You should go back to the regex in `enableTsDiagnostics` as we said above. And yeah you have to make it actually a regex with `RegEx(...)` or `new RegEx(...)`, sorry.

For the tests:
- duplicate the whole test case folder `tests/ts-diagnostics` into `tests/ts-diagnostics-regex` for example
- in that new test case, make 2 source files with 2 tests, and in package.json make an `enableTsDiagnostics` regex that matches only one of the source files and/or test files
- in the test file (`tests/__tests__/ts-diagnostics.spec.ts`):
  - duplicate the it block and in the new one replace `runJest('../ts-diagnostics', ['--no-cache'])` with `runJest('../ts-diagnostics-regex')` (disabling the cache should do the trick about the 2 error messages

lyy011lyy commented:
Hi @huafu :
Thanks for suggestion.
I think I didn't explain it very well. The reason that type error in Hello.ts is always shown is that when we preprocess Hello.test.ts, it will also do the diagnostic for Hello.ts (This will happen in preprocessing Hello.test.ts which will never be skipped).  And this is not only related to the test and it also happens to my project. So I think this is the actual behavior of our diagnostic.

So it has two steps in total:
1. Preprocess Hello.test.ts -> diagnose both (Hello.test.ts and Hello.ts)
2. Preprocess Hello.ts -> diagnose Hello.ts (Duplicated one)

Our change can only remove the second step,  but the diagnostic will always happen on both Hello.test.ts and Hello.ts.
That's the reason why I just change it back to your previous suggestion.

huafu commented:
@lyy011lyy then it's maybe another issue that needs to be fixed. Maybe you can pass the config to `runTsDiagnostics()` so that in `https://github.com/kulshekhar/ts-jest/blob/master/src/utils/run-ts-diagnostics.ts#L8-L18`, the diagnostics are filtered based on file path again. Also there you can create a global map to ensure a diagnostic is thrown only once per jest instance

lyy011lyy commented:
@huafu  Yes, I think we found a duplicated diagnostic. I see what you mean, Let me try it. 
Thanks for your time to help me :)

lyy011lyy commented:
Hi @huafu 
You are right, we are doing the diagnostic for all files in the run-ts-diagnostics:
it's  https://github.com/kulshekhar/ts-jest/blob/master/src/utils/run-ts-diagnostics.ts#L13
According to doc: https://github.com/Microsoft/TypeScript-wiki/blob/master/API-Breaking-Changes.md#program-interface-changes
If we don't path the second parameter: SourceFile, typescript will do the diagnostic for all files.

So I add it and now it will only diagnose the file we pass in. And it's dependencies will be controlled by the regex.
The tests are also added based on your suggestion.
Really appreciate your time to help :)

huafu commented:
@lyy011lyy thanks a lot for the work! You're missing one thing tho. You added the test case folder, but you need to add the test that will run jest in the test case folder and check the output:

> in the test file (`tests/__tests__/ts-diagnostics.spec.ts`), duplicate the `it(...)` block and in the new one replace `runJest('../ts-diagnostics', ['--no-cache'])` with `runJest('../ts-diagnostics-regex', ['--no-cache'])`

lyy011lyy commented:
@huafu Oh, sorry about that, I added it but forgot to add it to the commit....
I don't know how to run a single test, so I moved out all other test files and only leave ts-diagnostic.spec, so when I select files to commit, there are too many test files changed and it's missed.

lyy011lyy commented:
Hi @huafu   the test is added, could you please help to take a look and merge it ?
Thanks :)

huafu commented:
fixes #678

GeeWee commented:
If the CI tests pass, let's merge this in!

lyy011lyy commented:
@huafu  @GeeWee Thanks for merging this!
Can I ask when will this change be released? 
Our project needs this ts diagnostic feature for all tests, hopefully, we can get it asap.
Thanks! :)

kulshekhar commented:
@lyy011lyy if you can send a quick PR increasing the version patch in package.json, I'll merge that in and publish that right away

lyy011lyy commented:
Hi @kulshekhar :
The new pull request is created: https://github.com/kulshekhar/ts-jest/pull/683
Thanks for your help !

kulshekhar commented:
@lyy011lyy 23.1.4 has been published. Thanks for your contribution :)
